### PR TITLE
Invert set retain

### PR DESCRIPTION
**Pull request by vogel**
setRetain logic is now inverted - BucketManager checks if bucket should be retained at the only 2 places that did setRetain(true) before - in current HistoryArchiveState and in publish queue table in database.

This is much simpler solution.

Additionaly, I've moved bucket iterator classes to separate files.

### CODE REVIEW

**Comment by MonsieurNicolas** [comment]
I think that we'll include this in the next release to limit churn

**Comment by vogel** [comment]
Rebased

**Comment by MonsieurNicolas** [comment]
this needs to be rebased

**Comment by vogel** [comment]
Rebased. #1458 delayed until new implementation is done (with listing all files).

File: src/bucket/BucketInputIterator.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
```
Comment (MonsieurNicolas) [inline src/bucket/BucketInputIterator.cpp]: new files need to be added to Visual Studio project files

**Comment by vogel** [comment]
Added files to VS project

**Comment by MonsieurNicolas** [comment]
r+ 075b2892190d9db09d31796741a10358a563a3ba

**Pull request merged**